### PR TITLE
feat: add DelegatedController

### PR DESCRIPTION
**Pull request by lbertenasco**

### CODE REVIEW

File: contracts/controllers/DelegatedController.sol [hunk 1]
```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.5.16;
```
Comment (fubuloubu) [inline contracts/controllers/DelegatedController.sol:3]: Change to 0.5.17 please?

File: contracts/controllers/DelegatedController.sol [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    ) external {
+        require(msg.sender == governance, "!governance");
+        converters[_input][_output] = _converter;
+    }
+
+    function setStrategy(address _vault, address _strategy) external {
+        require(msg.sender == governance, "!governance");
+        address _current = strategies[_vault];
+        if (_current != address(0)) {
+            Strategy(_current).withdrawAll();
```
Comment (fubuloubu) [inline contracts/controllers/DelegatedController.sol:71]: General question: how does the capital get re-invested in the new strategy?

File: contracts/controllers/DelegatedController.sol [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        isStrategy[_strategy] = true;
+        vaults[_strategy] = _vault;
+        isVault[_vault] = true;
+    }
+
+    function want(address _vault) external view returns (address) {
+        return Strategy(strategies[_vault]).want();
+    }
+
+    function earn(address _vault, uint256 _amount) public {
```
Comment (fubuloubu) [inline contracts/controllers/DelegatedController.sol:83]: Anyone is allowed to up the capital a Strategy has? This seems bad (even though there's no direct theft happening)

File: contracts/controllers/DelegatedController.sol [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        (expected, ) = OneSplitAudit(onesplit).getExpectedReturn(_token, _want, _balance, parts, 0);
+    }
+
+    function claimInsurance(address _vault) external {
+        require(msg.sender == governance, "!governance");
+        DelegatedVault(_vault).claimInsurance();
+    }
+
+    // Only allows to withdraw non-core strategy tokens ~ this is over and above normal yield
+    function delegatedHarvest(address _strategy, uint256 parts) external {
```
Comment (fubuloubu) [inline contracts/controllers/DelegatedController.sol:127]: This is... a lot to digest. Will have to take a second look to see if I fully understand it.

File: contracts/controllers/DelegatedController.sol [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                _amount = _after.sub(_before);
+                uint256 _reward = _amount.mul(split).div(max);
+                IERC20(_want).safeTransfer(vaults[_strategy], _amount.sub(_reward));
+                IERC20(_want).safeTransfer(rewards, _reward);
+            }
+        }
+    }
+
+    // Only allows to withdraw non-core strategy tokens ~ this is over and above normal yield
+    function harvest(
```
Comment (fubuloubu) [inline contracts/controllers/DelegatedController.sol:154]: Need to understand the difference between this and `delegatedHarvest`, and whether there's an oppurtunity for de-duplication

File: contracts/controllers/DelegatedController.sol (hunk 3 above)
Comment (milkyklim) [reply contracts/controllers/DelegatedController.sol:83]: Seems like intentional design. Incentives are not there yet but idea is to have some type of "keepers" calling `earn` when necessary.

File: contracts/controllers/DelegatedController.sol (hunk 2 above)
Comment (storming0x) [reply contracts/controllers/DelegatedController.sol:71]: Not 100% sure,  but only way is doing it via `earn` method on lines `86,87` but you have to match the strategy/vault key pair you added here

File: contracts/controllers/DelegatedController.sol (hunk 3 above)
Comment (storming0x) [reply contracts/controllers/DelegatedController.sol:83]: Agree, also some checks on line 85 like `require(strategy !== 0x0,"!bad_strategy")`

And a check that the amount and balance are correct? not sure.

File: interfaces/yearn/DelegatedVault.sol [hunk 6]
```diff
@@ -0,0 +1,17 @@
+pragma solidity ^0.5.16;
```
Comment (storming0x) [inline interfaces/yearn/DelegatedVault.sol:1]: should be all `0.5.17`?

**Review by milkyklim — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]
If this matches on-chain (ignoring comments/interface changes) then g2g

**Comment by milkyklim** [comment]
@lbertenasco can you confirm that this one is fine? I'll create another PR for solc 0.5.17 update.

**Pull request merged**